From 556eb42758f326418f9d4bf0ebf02c0344b5f638 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 10:12:53 -0400
Subject: [fs] ext4: fix COLLAPSE RANGE test for bigalloc file systems

Message-id: <1414059176-6044-30-git-send-email-lczerner@redhat.com>
Patchwork-id: 98292
O-Subject: [RHEL 7.1 PATCH 29/32] ext4: fix COLLAPSE RANGE test for bigalloc file systems
Bugzilla: 1150171

From: Namjae Jeon <namjae.jeon@samsung.com>

BZ 1150171
https://bugzilla.redhat.com/show_bug.cgi?id=1150171

BREW 8149772
https://brewweb.devel.redhat.com/taskinfo?taskID=8149772

Upstream commit ee98fa3a8b148a234600a20f7cdc2b4b37f38083

 Blocks in collapse range should be collapsed per cluster unit when
 bigalloc is enable. If bigalloc is not enable, EXT4_CLUSTER_SIZE will
 be same with EXT4_BLOCK_SIZE.

 With this bug fixed, patch enables COLLAPSE_RANGE for bigalloc, which
 fixes a large number of xfstest failures which use fsx.

 Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
 Signed-off-by: Ashish Sangwan <a.sangwan@samsung.com>
 Signed-off-by: Theodore Ts'o <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index b063345..9aa6bc6 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -5395,16 +5395,13 @@ int ext4_collapse_range(struct inode *inode, loff_t offset, loff_t len)
  int ret;
 
  /* Collapse range works only on fs block size aligned offsets. */
- if (offset & (EXT4_BLOCK_SIZE(sb) - 1) ||
-     len & (EXT4_BLOCK_SIZE(sb) - 1))
+ if (offset & (EXT4_CLUSTER_SIZE(sb) - 1) ||
+     len & (EXT4_CLUSTER_SIZE(sb) - 1))
   return -EINVAL;
 
  if (!S_ISREG(inode->i_mode))
   return -EINVAL;
 
- if (EXT4_SB(inode->i_sb)->s_cluster_ratio > 1)
-  return -EOPNOTSUPP;
-
  trace_ext4_collapse_range(inode, offset, len);
 
  punch_start = offset >> EXT4_BLOCK_SIZE_BITS(sb);
-- 
1.7.1